pkg/ingester/ingester.go

```diff
@@ -518,6 +521,15 @@ func (i *Ingester) starting(ctx context.Context) (err error) {
 		return errors.Wrap(err, "opening existing TSDBs")
 	}
 
+	// When ingest storage is enabled, we have to make sure that reader catches up replaying the partition
+	// BEFORE the ingester ring lifecycler is started, because once the ingester ring lifecycler will start
+	// it will switch the ingester state to ACTIVE.
+	if i.ingestReader != nil {
+		if err := services.StartAndAwaitRunning(ctx, i.ingestReader); err != nil {
+			return errors.Wrap(err, "failed to start partition reader")
+		}
+	}
```

Comment:
> doing this here means that some ingester metrics aren't updated and owned series and compaction are off. Can this cause problems when the ingester has been down for a long time (i.e. need to do compaction while catching up with a long backlog?)

This is very good feedback. I need to think more how to solve it. However I don't think we can move this logic after switching an ingester to ACTIVE in the ring (because of how we designed partitions lifecycle), and changing ingesters lifecycler looks a too broad work for now.

I'm wondering if we should focus on the normal restart case, and work on the "restart after a long outage" case later on, to have something to move on in the meanwhile.

Reply:
After agreeing a design with @pstibrany , I've reworked this PR to better organise how internal services are started in `Ingester.starting()` so that Head compaction, owned series, metrics updater and metadata purger are all running before we start replaying the partition, while keeping the ingester lifecycler started after we replayed the partition.